Teach typedConstructorPattern to ignore non-redefining unapplies

dwijnand opened the pull request:
In pos/t12250.scala Foo defines an unapply which doesn't redefine the
synthetic one.  Thus unapply is overloaded.  In the pattern being typed,
the synthetic one is indeed the one seeked, so despite being overloaded,
we should convertToCaseConstructor.

Fixes https://github.com/scala/bug/issues/12250

dwijnand committed:
Teach typedConstructorPattern to ignore non-redefining unapplies

In pos/t12250.scala Foo defines an unapply which doesn't redefine the
synthetic one.  Thus unapply is overloaded.  In the pattern being typed,
the synthetic one is indeed the one seeked, so despite being overloaded,
we should convertToCaseConstructor.

lrytz commented:
🤔
```
$> cat Test.scala
object T {
  trait A extends Any { def f: Object = "a" }
  implicit class C(private val x: String) extends AnyVal with A {
    @annotation.inheritSignature override def f: String = "c"
    def g: String = "c"
  }
  def main(args: Array[String]): Unit = {
    println("".f)
    println("".g)
  }
}
$> scpr 9462 A.scala
A.scala:7: warning: match may not be exhaustive.
It would fail on the following input: Foo(_)
    id match {
    ^
1 warning
```

dwijnand commented:
Oh, I took out the -Werror I had initially... because I somehow convinced myself pos tests fail on warnings... 🤯

dwijnand commented:
Fixed my boolean algebra mistake.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala:
```diff
@@ -97,9 +97,9 @@ trait PatternTypers {
-      else if (canElide && caseClass.isCase && !member.isOverloaded && (member == NoSymbol || member.isSynthetic))
+      // A case class may have its own custom unapply (so non-synthetic) pos/t11252.scala
+      // A case class may have its own custom unapply unrelated to deconstructing pos/t12250.scala
+      else if (canElide && caseClass.isCase && (!member.isOverloaded && (member == NoSymbol || member.isSynthetic) || member.alternatives.exists(_.isSynthetic) && caseClass.tpe =:= pt))
```
I don't know my way around here that well; IIUC this assumes that the synthetic `unapply` overload is the one selected - could it be that overloading resolution picks a different one, would it be fine to go in`convertToCaseConstructor`?

dwijnand replied:
I think there is only ever one synthetic unapply for a case class (`caseClass.isCase`).  And your understanding is correct, except it's not really "selected": the presence of the case class unapply means that using its constructor is the right course of action.

Now that I think about it, I should review the case-class-23 situation.  Basically it never had an unapply and we rewrote it to use its constructor.  What happens in the various situations of redefining its unapply or defining an un-related unapply? todo += 1

dwijnand replied:
OK, the situation is there's an issue (IMO) but thankfully my change here didn't affected it.

dwijnand committed:
Test size=23 unrelated/redefined unapplies

Just recording the current behaviour, to confirm the next commit doesn't
regress them.

dwijnand commented on src/compiler/scala/tools/nsc/typechecker/PatternTypers.scala:
```diff
@@ -97,9 +97,9 @@ trait PatternTypers {
-      else if (canElide && caseClass.isCase && !member.isOverloaded && (member == NoSymbol || member.isSynthetic))
+      // A case class may have its own custom unapply (so non-synthetic) pos/t11252.scala
+      // A case class may have its own custom unapply unrelated to deconstructing pos/t12250.scala
+      else if (canElide && caseClass.isCase && (!member.isOverloaded && (member == NoSymbol || member.isSynthetic) || member.alternatives.exists(_.isSynthetic) && caseClass.tpe =:= pt))
```
Talking about this diff in passing @adriaanm mentioned `isApplicable` in Infer, which is already `import infer._`-ed above, so I should see if I can use that over `=:=`.

dwijnand replied:
Turned out not to be necessary. `=:=` is right, and I wrote the super-type/sub-type variations to prove that it's right (or at least convinced myself it's right... confirmation bias...).

I also reviewed how user-defined unapplies (and applies) are determined (#5816) and it's using Type#matchesType.  But, again, `=:=` looks to be correct while `<:<` isn't, which was my original thought.

dwijnand commented:
Seth was ready to merge, but I wanted to review this detail, so it can be merged now once it's green.
